File: src/webots/launcher/launcher.c
/*
 * Copyright 1996-2023 Cyberbotics Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Description:  This is the source code of webots.exe and webotsw.exe, the binary launchers of webots-bin.exe (Windows only).
//               It adds paths to the PATH environment variable to ensure that the correct libraries will be found and starts
//               webots-bin.exe, passing all the command line arguments, wait until the completion of webots-bin.exe and
//               returns the exit status code of webots-bin.exe.
//               The main advantage of a binary launcher over a batch launcher (webots.bat) is that the binary launcher
//               doesn't open a DOS cmd.exe console in the background, whereas this is unavoidable when running a batch file.
//               webotsw.exe is a windows application (to be started from the icon or menu).
//               webots.exe is DOS application (to be started from a DOS command prompt or a script).
//               Starting webots.exe from the icon is fine, but will open a DOS command prompt in the background.
//               (a similar naming convention is used for python.exe / pythonw.exe, java.exe / javaw.exe, etc.)

#include <shlwapi.h>
#include <stdio.h>
#include <windows.h>

static int fail(const char *function, const char *info) {
  DWORD e = GetLastError();
  if (e) {
    LPSTR m = NULL;
    FormatMessage(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS, NULL, e,
                  MAKELANGID(LANG_ENGLISH, SUBLANG_ENGLISH_US), (LPSTR)&m, 0, NULL);
    char message[1024];
    const char *lf = info ? "\n" : "";
    const char *i = info ? info : "";
    // cppcheck-suppress nullPointer
    snprintf(message, sizeof(message), "%s failed with error %lu.\n%s%s%s", function, e, m, i, lf);
    LocalFree(m);
#ifdef WEBOTSW
    MessageBox(NULL, message, "Webots launcher error", MB_ICONERROR | MB_OK);
#else
    fprintf(stderr, message);
#endif
  } else
    fprintf(stderr, "%s failed with no error.\n", function);
  exit(e);
}

int main(int argc, char *argv[]) {
  // We retrieve the command line in wchar_t from the Windows system.
  const int LENGTH = 4096;
  wchar_t *module_path = malloc(LENGTH * sizeof(wchar_t));
  if (!GetModuleFileNameW(NULL, module_path, LENGTH))
    fail("GetModuleFileNameW", 0);
  const int l = wcslen(module_path)
#ifdef WEBOTSW
                - 1  // webotsw.exe (we need to remove the final 'w')
#endif
    ;
  wchar_t *command_line = malloc(LENGTH * sizeof(wchar_t));
  // In order to launch Webots, we simply need to replace 'webotsw.exe'/'webots.exe' with 'webots-bin.exe'
  const int index = l - 4;  // don't copy the ".exe"
  wcsncpy(command_line, module_path, index);
  command_line[index] = L'\0';
  wcscat(command_line, L"-bin.exe");
  const wchar_t *arguments = PathGetArgsW(GetCommandLineW());
  if (arguments && arguments[0] != L'\0') {
    wcscat(command_line, L" ");
    wcscat(command_line, arguments);
  }
  // add "WEBOTS_HOME/msys64/mingw64/bin", "WEBOTS_HOME/msys64/mingw64/bin/cpp" and "WEBOTS_HOME/msys64/usr/bin" to the PATH
  // environment variable
  wchar_t *old_path = malloc(LENGTH * sizeof(wchar_t));
  wchar_t *new_path = malloc(LENGTH * sizeof(wchar_t));

  wcscpy(new_path, module_path);
  new_path[l - 11] = ';';  // removes "\webots.exe" or "\webotsw.exe"
  wcscpy(&new_path[l - 10], module_path);
  new_path[2 * l - 21] = '\0';
  wcscat(new_path, L"\\cpp;");
  wcscat(new_path, module_path);
  free(module_path);
  new_path[3 * l - 38] = '\0';
  wcscat(new_path, L"usr\\bin;");
  if (!GetEnvironmentVariableW(L"PATH", old_path, LENGTH))
    fail("GetEnvironmentVariableW", "PATH");
  wcscat(new_path, old_path);
  free(old_path);
  if (!SetEnvironmentVariableW(L"PATH", new_path))
    fail("SetEnvironmentVariableW", "PATH");
  free(new_path);
  if (!SetEnvironmentVariableW(L"QT_ENABLE_HIGHDPI_SCALING", L"1"))
    fail("SetEnvironmentVariableW", "QT_ENABLE_HIGHDPI_SCALING=1");

  // if set, we need to remove this environment variable set by Qt5 which conflicts with Qt6
  SetEnvironmentVariableW(L"QT_QPA_PLATFORM_PLUGIN_PATH", NULL);

  // start the webots-bin.exe process, wait for completion and return exit code
  STARTUPINFOW info = {sizeof(info)};
  PROCESS_INFORMATION process_info;

  while (1) {
    if (!CreateProcessW(NULL, command_line, NULL, NULL, TRUE, 0, NULL, NULL, &info, &process_info))
      fail("CreateProcess", "Cannot launch Webots binary");

    // webots-bin.exe should be killed whenever its parent (webots.exe or webotsw.exe) terminates.
    HANDLE job = CreateJobObject(NULL, NULL);
    JOBOBJECT_EXTENDED_LIMIT_INFORMATION jeli = {0};
    jeli.BasicLimitInformation.LimitFlags = JOB_OBJECT_LIMIT_KILL_ON_JOB_CLOSE;
    if (!SetInformationJobObject(job, JobObjectExtendedLimitInformation, &jeli, sizeof(jeli)))
      fail("SetInformationJobObject", 0);
    if (!AssignProcessToJobObject(job, process_info.hProcess))
      fail("AssignProcessToJobObject", 0);

    // wait for webots-bin.exe to terminate
    WaitForSingleObject(process_info.hProcess, INFINITE);  // return zero in case of success
    DWORD exit_code;
    if (!GetExitCodeProcess(process_info.hProcess, &exit_code))
      fail("GetExitCodeProcess", 0);
    if (!CloseHandle(process_info.hProcess))
      fail("CloseHandle", 0);
    if (!CloseHandle(process_info.hThread))
      fail("CloseHandle", 0);
    if (exit_code != 3030)  // special return code to restart Webots, see WbGuiApplication.cpp
      return exit_code;
  }
  free(command_line);
  return 0;
}

File: projects/samples/howto/docker/launcher.py
# Copyright 1996-2023 Cyberbotics Ltd.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
import subprocess
import sys


def run(command):
    try:
        process = subprocess.Popen(command.split(),
                                   stdout=subprocess.PIPE,
                                   stderr=subprocess.STDOUT,
                                   bufsize=1, universal_newlines=True)
    except Exception:
        print(f"Error: Unable to start {command}")
        quit()
    while process.poll() is None:
        line = process.stdout.readline().rstrip()
        if line:
            print(line)


if sys.platform != 'linux':
    sys.exit('This script runs only on Linux, not on ' + sys.platform)

controllers = {
    "MyBot": "controllers/camera/camera"
}
port = 1234
docker_image = 'cyberbotics/webots:R2023b-ubuntu22.04'  # this should correspond to the current version of Webots

subprocess.run(['xhost', '+local:root'])
with open('simulation/.env', 'w+') as env_file:
    env_file.write(f'IMAGE={docker_image}\n')
    env_file.write(f'PORT={port}\n')
    env_file.write('ROBOT_NAME_1=MyBot\n')
    env_file.write('WORLD=/webots_project/worlds/camera.wbt\n')

command = 'docker compose -f simulation/docker-compose-webots.yml up --build --no-color'
try:
    webots_process = subprocess.Popen(command.split(),
                                      stdout=subprocess.PIPE,
                                      stderr=subprocess.STDOUT,
                                      bufsize=1, universal_newlines=True)
except Exception:
    print(f"error: Unable to start docker compose: {command}")
    quit()
controller_process = None
while webots_process.poll() is None:
    line = webots_process.stdout.readline().rstrip()
    if line.startswith('simulation-webots-1  | '):  # output of the first docker container
        line = line[23:]
        if line.startswith('ipc://'):
            name = line[line.rfind('/') + 1:]
            controller = controllers[name] if name in controllers else ''
            if not controller:
                print('controller for robot "' + name + '" not found, skipping...')
                continue
            folder = os.path.split(controller)[0]
            command = ('docker build -t controller '
                       f'--build-arg WEBOTS_DEFAULT_IMAGE={docker_image} '
                       '--build-arg MAKE=1 '
                       f'{folder}')
            run(command)
            command = (f'docker run -e WEBOTS_CONTROLLER_URL=ipc://{port}/{name} '
                       f'-v tmp-{port}-{name}:/tmp/webots-{port}/ipc/{name} controller /webots_project/{controller}')
            subprocess.Popen(command.split())  # launch in the background
    elif line:
        print(line)  # docker compose output
    if controller_process:
        while controller_process.poll() is None:
            line = controller_process.stdout.readline().rstrip()
            if line:
                print('controller: ' + line)
os.remove('simulation/.env')
